File: src/Native/GenericReflection.cpp
#include "GenericReflection.h"
#include "VariableReflection.h"
#include "TypeReflection.h"
#include <stdexcept>

Native::GenericReflection::GenericReflection(void* native)
{
    if (!native) throw std::invalid_argument("Native pointer cannot be null");

	m_native = (slang::GenericReflection*)native;
}

Native::GenericReflection::~GenericReflection()
{
}

slang::GenericReflection* Native::GenericReflection::getNative()
{
    return m_native;
}

//Native::DeclReflection* Native::GenericReflection::asDecl()
//{
//    return (DeclReflection*)spReflectionGeneric_asDecl((SlangReflectionGeneric*)this);
//}

char const* Native::GenericReflection::getName() 
{
    return m_native->getName();
}

unsigned int Native::GenericReflection::getTypeParameterCount()
{
    return m_native->getTypeParameterCount();
}

Native::VariableReflection* Native::GenericReflection::getTypeParameter(unsigned index)
{
    slang::VariableReflection* nativeTypeParameter = m_native->getTypeParameter(index);
    return nativeTypeParameter ? new Native::VariableReflection(nativeTypeParameter) : nullptr;
}

unsigned int Native::GenericReflection::getValueParameterCount()
{
    return m_native->getValueParameterCount();
}

Native::VariableReflection* Native::GenericReflection::getValueParameter(unsigned index)
{
    slang::VariableReflection* nativeValueParameter = m_native->getValueParameter(index);
    return nativeValueParameter ? new Native::VariableReflection(nativeValueParameter) : nullptr;
}

unsigned int Native::GenericReflection::getTypeParameterConstraintCount(VariableReflection* typeParam)
{
    return m_native->getTypeParameterConstraintCount((slang::VariableReflection*)typeParam->getNative());
}

Native::TypeReflection* Native::GenericReflection::getTypeParameterConstraintType(VariableReflection* typeParam, unsigned index)
{
    slang::TypeReflection* nativeResult = m_native->getTypeParameterConstraintType((slang::VariableReflection*)typeParam->getNative(), index);
    return nativeResult ? new TypeReflection(nativeResult) : nullptr;
}

//Native::DeclReflection* Native::GenericReflection::getInnerDecl()
//{
//    return (DeclReflection*)spReflectionGeneric_GetInnerDecl((SlangReflectionGeneric*)this);
//}

Native::DeclKind Native::GenericReflection::getInnerKind()
{
    return (Native::DeclKind)m_native->getInnerKind();
}

Native::GenericReflection* Native::GenericReflection::getOuterGenericContainer()
{
    slang::GenericReflection* outerContainerPtr = m_native->getOuterGenericContainer();
    return outerContainerPtr ? new GenericReflection(outerContainerPtr) : nullptr;
}

Native::TypeReflection* Native::GenericReflection::getConcreteType(VariableReflection* typeParam)
{
    slang::TypeReflection* nativeResult = m_native->getConcreteType((slang::VariableReflection*)typeParam->getNative());
    return nativeResult ? new TypeReflection(nativeResult) : nullptr;
}

int64_t Native::GenericReflection::getConcreteIntVal(VariableReflection* valueParam)
{
	return m_native->getConcreteIntVal((slang::VariableReflection*)valueParam->getNative());
}

Native::GenericReflection* Native::GenericReflection::applySpecializations(GenericReflection* genRef)
{
    slang::GenericReflection* nativeResult = m_native->applySpecializations((slang::GenericReflection*)genRef->getNative());
    return nativeResult ? new Native::GenericReflection(nativeResult) : nullptr;
}

File: src/Slang.Sdk/Layer2-Binding/Reflection-API/GenericReflection.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Slang.Sdk.Interop;
using static Slang.Sdk.Interop.Utilities;

namespace Slang.Sdk.Binding
{
    internal unsafe class GenericReflection : Reflection
    {
        internal override Reflection? Parent { get; }
        internal override GenericReflectionHandle Handle { get; }

        internal GenericReflection(Reflection parent, GenericReflectionHandle handle)
        {
            Parent = parent;
            Handle = handle;
        }

        internal string GetName()
        {
            string? error = null;
            return Call(() => StrongInterop.GenericReflection.GetName(Handle, out error), () => error);
        }

        internal uint GetTypeParameterCount()
        {
            string? error = null;
            return Call(() => StrongInterop.GenericReflection.GetTypeParameterCount(Handle, out error), () => error);
        }

        internal VariableReflection GetTypeParameter(uint index)
        {
            string? error = null;
            return new VariableReflection(this, Call(() => StrongInterop.GenericReflection.GetTypeParameter(Handle, index, out error), () => error));
        }

        internal uint GetValueParameterCount()
        {
            string? error = null;
            return Call(() => StrongInterop.GenericReflection.GetValueParameterCount(Handle, out error), () => error);
        }

        internal VariableReflection GetValueParameter(uint index)
        {
            string? error = null;
            return new VariableReflection(this, Call(() => StrongInterop.GenericReflection.GetValueParameter(Handle, index, out error), () => error));
        }

        internal uint GetTypeParameterConstraintCount(TypeParameterReflection typeParam)
        {
            string? error = null;
            return Call(() => StrongInterop.GenericReflection.GetTypeParameterConstraintCount(Handle, typeParam.Handle, out error), () => error);
        }

        internal TypeReflection GetTypeParameterConstraintType(TypeParameterReflection typeParam, uint index)
        {
            string? error = null;
            return new TypeReflection(this, Call(() => StrongInterop.GenericReflection.GetTypeParameterConstraintType(Handle, typeParam.Handle, index, out error), () => error));
        }

        internal int GetInnerKind()
        {
            string? error = null;
            return Call(() => StrongInterop.GenericReflection.GetInnerKind(Handle, out error), () => error);
        }

        internal GenericReflection GetOuterGenericContainer()
        {
            string? error = null;
            return new GenericReflection(this, Call(() => StrongInterop.GenericReflection.GetOuterGenericContainer(Handle, out error), () => error));
        }

        internal TypeReflection GetConcreteType(TypeParameterReflection typeParam)
        {
            string? error = null;
            return new TypeReflection(this, Call(() => StrongInterop.GenericReflection.GetConcreteType(Handle, typeParam.Handle, out error), () => error));
        }

        internal long GetConcreteIntVal(VariableReflection valueParam)
        {
            string? error = null;
            return Call(() =>
            {
                long value;
                StrongInterop.GenericReflection.GetConcreteIntVal(Handle, valueParam.Handle, out value, out error);
                return value;
            }, () => error);
        }

        internal GenericReflection ApplySpecializations(GenericReflection genRef)
        {
            string? error = null;
            return new GenericReflection(this, Call(() => StrongInterop.GenericReflection.ApplySpecializations(Handle, genRef.Handle, out error), () => error));
        }
    }
}
